ui/src/components/Canvas.tsx

```diff
@@ -1145,46 +506,267 @@ export function Canvas() {
+  let keys = new Set(nodesMap.keys());
+  let nodesMap2 = new Map<string, Node>();
+  nodes.forEach((node) => nodesMap2.set(node.id, node));
+  let keys2 = new Set(nodesMap2.keys());
+  if (keys.size !== keys2.size) {
+    console.error("keys are not the same", keys, keys2);
+    return false;
+  }
+  for (let i = 0; i < keys.size; i++) {
+    if (keys[i] !== keys2[i]) {
+      console.error("keys are not the same", keys, keys2);
+      return false;
+    }
+  }
+  // verify the values
+  keys.forEach((key) => {
+    let node1 = nodesMap.get(key);
+    let node2 = nodesMap2.get(key);
+    if (!node1) {
+      console.error("node1 is undefined");
+      return false;
+    }
+    if (!node2) {
+      console.error("node2 is undefined");
+      return false;
+    }
+    if (node1.id !== node2.id) {
+      console.error("node id are not the same", node1.id, node2.id);
+      return false;
+    }
+    if (node1.parentNode !== node2.parentNode) {
+      console.error(
+        "node parent are not the same",
+        node1.parentNode,
+        node2.parentNode
+      );
+      return false;
+    }
+    if (node1.position.x !== node2.position.x) {
+      console.error(
+        "node x are not the same",
+        node1.position.x,
+        node2.position.x
+      );
+      return false;
+    }
+    if (node1.position.y !== node2.position.y) {
+      console.error(
+        "node y are not the same",
+        node1.position.y,
+        node2.position.y
+      );
+      return false;
+    }
+  });
+  return true;
+}
+
+function useInitNodes({ triggerUpdate }) {
+  const store = useContext(RepoContext)!;
+  const getPod = useStore(store, (state) => state.getPod);
+  const nodesMap = useStore(store, (state) => state.ydoc.getMap<Node>("pods"));
+  const getId2children = useStore(store, (state) => state.getId2children);
+  const provider = useStore(store, (state) => state.provider);
+  const [loading, setLoading] = useState(true);
+  useEffect(() => {
+    const init = () => {
+      let nodes = store2nodes("ROOT", -1, { getId2children, getPod });
+      // Verify that the nodes are the same as the remote database
+      if (nodesMap.size !== nodes.length) {
+        console.info(
+          "The yjs server is empty but database is not. Initializing the yjs server."
+        );
+        nodes.forEach((node) => {
+          if (!nodesMap.has(node.id)) {
+            nodesMap.set(node.id, node);
+          }
+        });
+      }
+      let isConsistent = verifyConsistency(nodes, nodesMap);
+      if (!isConsistent) {
+        console.warn(
+          "The yjs server is not consistent with the database. Resetting the yjs server"
+        );
+        // throw new Error("Inconsistent state");
+        nodes.forEach((node) => {
+          if (!nodesMap.has(node.id)) {
+            nodesMap.set(node.id, node);
+          }
+        });
+      }
+      // NOTE we have to trigger an update here, otherwise the nodes are not
+      // rendered.
+      triggerUpdate();
```

Comment:
This is actually where I use `triggerUpdate`. Without this, the canvas will not show anything.